pikachewie/publisher.py

```diff
@@ -51,6 +51,20 @@ def publish(self, exchange, routing_key, body, properties=None):
                 body=body,
             )
 
+    def process_data_events(self, time_limit=0):
+        """Calls process_data_events on the current connection (if there is
+        currently one open).
+        """
+        if self._channel:
+            try:
+                self.channel.connection.process_data_events(time_limit=time_limit)
+            except self.retry_on_exceptions as exc:
+                log.warn('Cannot process data events')
+
+                # If there's a problem, we'll discard the channel and reconnect
+                # it on demand.
+                self._channel = None
```

Comment:
I was hoping you wouldn't have an issue with that inconsistency. :) I'm conflating channel and connection to a certain extent, mainly because the module doesn't distinguish between the two either.

Take the `publish` method. If it encounters an closed channel or connection, then it removes the reference to channel. This forces a new channel to be created, but it does that using the line:

    self._channel = self.broker.connect(blocking=True).channel()

Looking at the documentation of the broker module for the connect method, it says `Return a new connection to this broker`. So the publisher currently only creates a connection when it is also creating a channel, and disregards both when something goes wrong. Since we don't currently care about the connection outside the context of the channel, I thought I was being consistent in this regard.

As we're losing a reference to the connection after an error occurs, I didn't see the point in forcing a new connection to be created in this method to perform heartbeat related tasks.

If this doesn't make sense, or you disagree - feel free to ping me on the usual channels. ;)